src/server/launchpad/client.js

```diff
@@ -0,0 +1,347 @@
+   * @param size - How many entries to serve.
+   * @returns {Promise<Collection, ResourceError>}
+   */
+  lp_slice(start, size) {
+    return this.lp_client.get(this.uri, { start: start, size: size });
+  }
+}
+
+/** A single object from the Launchpad web service. */
+export class Entry extends Resource {
+  constructor(client, representation, uri) {
+    super();
+    this.lp_client = client;
+    this.uri = uri;
+    this.lp_attributes = {};
+    this.dirty_attributes = [];
+
+    // Copy the representation keys into our own set of attributes, and add
+    // an attribute-change event listener for caching purposes.
+    for (const key of Object.keys(representation)) {
+      this.lp_attributes[key] = representation[key];
+      Object.defineProperty(this, key, {
+        configurable: false,
+        enumerable: true,
+        get: () => { return this.lp_attributes[key]; },
+        set: value => {
+          if (value !== this.lp_attributes[key]) {
+            this.dirty_attributes.push(key);
+          }
+          this.lp_attributes[key] = value;
+        }
+      });
+    }
+  }
+
+  /** Write modifications to this entry back to the web service. */
+  lp_save(config) {
+    let representation = {};
+    for (const attribute of this.dirty_attributes) {
+      representation[attribute] = this[attribute];
+    }
+    let headers = {};
+    if (this['http_etag'] !== undefined) {
+      headers['If-Match'] = this['http_etag'];
+    }
+    const uri = normalizeURI(this.lp_client.base_uri, this['self_link']);
+    return this.lp_client.patch(uri, representation, config, headers)
+      .then(() => { this.dirty_attributes = []; });
+  }
+}
+
+/** A client that makes HTTP requests to Launchpad's web service. */
+export class Launchpad {
+  constructor(base_uri, consumer_key, token_key, token_secret) {
+    this.base_uri = base_uri;
+    this.consumer_key = consumer_key;
+    this.token_key = token_key;
+    this.token_secret = token_secret;
+  }
+
+  makeHeaders() {
+    let headers = {};
+    if (this.consumer_key !== undefined &&
+        this.token_key !== undefined && this.token_secret !== undefined) {
+      // We leave a lot of the parameters null or empty here because
+      // Launchpad uses the PLAINTEXT method which really doesn't care.
+      const oauth = new OAuth(
+        null, null, this.consumer_key, '', '1.0', null, 'PLAINTEXT');
+      headers['Authorization'] = oauth.authHeader(
+        '', this.token_key, this.token_secret);
+    }
+    return headers;
+  }
+
+  /**
+   * Get the current state of a resource.
+   * @returns {Promise<Resource|Object, ResourceError>}
+   */
+  get(uri, config) {
+    if (config === undefined) {
+      config = {};
+    }
+    const start = config.start;
+    const size = config.size;
+    const headers = {
+      ...this.makeHeaders(),
+      'Accept': config.accept || 'application/json'
+    };
+    let parameters = { ...config.parameters };
```

Comment:
Just a detail, but `parameters` can be a const here it seems. As long as we don't do any other `parameters = ...` assignment parameters object can be changed with `parameters.something = ...` or `parameters['something'] = ...`.